src/librustc_mir/transform/qualify_consts.rs

```diff
@@ -968,12 +1004,42 @@ impl<'a, 'tcx> Checker<'a, 'tcx> {
                     bb = target;
                 }
                 _ => {
-                    self.not_const();
+                    self.not_const(ops::Loop);
+                    validator.check_op(ops::Loop);
                     break;
                 }
             }
         }
 
+        // The new validation pass should agree with the old when running on simple const bodies
+        // (e.g. no `if` or `loop`).
+        if !use_new_validator {
+            let mut new_errors = validator.take_errors();
+
+            // FIXME: each checker sometimes emits the same error with the same span twice in a row.
+            self.errors.dedup();
+            new_errors.dedup();
+
+            if self.errors != new_errors {
+                error!("old validator: {:?}", self.errors);
+                error!("new validator: {:?}", new_errors);
+
+                // ICE on nightly if the validators do not emit exactly the same errors.
+                // Users can supress this panic with an unstable compiler flag (hopefully after
+                // filing an issue).
+                let opts = &self.tcx.sess.opts;
+                let trigger_ice = opts.unstable_features.is_nightly_build()
```

Comment:
hey @rust-lang/compiler you may have thoughts on this trick. We have two (we think) equal algorithms that are checked against each other. If they are not equal, we ICE on nightly only, and there's a flag to skip the ICE.

Implementation wise the PR lgtm, so I think we should merge and see how it fares. It's very easy to turn the ICEing off if it gives us any trouble.

Reply:
sounds good to me :)